File: 트라이/개미굴_14725.java
package 트라이;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.StringTokenizer;

public class 개미굴_14725 {
	static int N;
	static Trie root;
	static int endCount;
	static int depth = 0;
	static StringBuilder answer = new StringBuilder();

	public static void main(String[] args) throws Exception {
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	    StringTokenizer st;

		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		root = new Trie();

		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());

			int len = Integer.parseInt(st.nextToken());
			String[] words = new String[len];
			for (int j = 0; j < len; j++) {
				words[j] = st.nextToken();
			}
			makeTrie(words);
		}

		makeVisualization(-1, root);
		System.out.print(answer);
	}

	private static void makeVisualization(int depth, Trie cur) {
		cur.getChild().sort(Comparator.comparing(N -> N.name));

		if (depth != -1) {
			for (int i = 0; i < depth; i++) {
				answer.append("--");
			}
			answer.append(cur.name).append("\n");
		}

		for (Trie child : cur.getChild()) {
			makeVisualization(depth + 1, child);
		}
	}

	private static void makeTrie(String[] words) {
		Trie cur = root;
		for (String name : words) {
			int idx = -1;

			for (int i = 0; i < cur.getChild().size(); i++) {
				if (cur.getChild().get(i).name.equals(name)) {
					idx = i;
					break;
				}
			}

			if (idx == -1) {
				Trie child = new Trie();
				child.name = name;
				cur.getChild().add(child);
				cur = cur.getChild().get(cur.getChild().size() - 1);
			} else {
				cur = cur.getChild().get(idx);
			}
		}
	}

	private static class Trie {
		String name = "";
		ArrayList<Trie> child = new ArrayList<>();

		public boolean hasChild() {
			return child != null;
		}

		public ArrayList<Trie> getChild() {
			return child;
		}
	}
}

File: 코틀린_코테스터디/week9/개미굴_14725.kt
package 코틀린_코테스터디.week9

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import java.util.function.Function


class 개미굴_14725 {
    private var N = 0
    private var root: Trie = Trie()
    private var answer = StringBuilder()

    fun solution() {
        val br = BufferedReader(InputStreamReader(System.`in`))
        var st: StringTokenizer
        st = StringTokenizer(br.readLine())
        N = st.nextToken().toInt()
        root = Trie()

        for (i in 0 until N) {
            st = StringTokenizer(br.readLine())
            val len = st.nextToken().toInt()
            val words = mutableListOf<String>()
            for (j in 0 until len) {
                words[j] = st.nextToken()
            }
            makeTrie(words)
        }
        makeVisualization(-1, root)
        print(answer)
    }

    private fun makeVisualization(depth: Int, cur: Trie) {
        cur.child.sortBy { it.name }
        if (depth != -1) {
            for (i in 0 until depth) {
                answer.append("--")
            }
            answer.append(cur.name).append("\n")
        }
        for (child in cur.child) {
            makeVisualization(depth + 1, child)
        }
    }

    private fun makeTrie(words: List<String>) {
        var cur = root
        for (name in words) {
            var idx = -1
            if (cur.child.isNotEmpty()) {
                for (i in cur.child.indices) {
                    if (cur.child[i].name == name) {
                        idx = i
                        break
                    }
                }
            }
            if (idx == -1) {
                val child = Trie()
                child.name = name
                cur.child.add(child)
                cur = cur.child[cur.child.size - 1]
            } else {
                cur = cur.child[idx]
            }
        }
    }

    private class Trie {
        var name: String = ""
        var child: MutableList<Trie> = mutableListOf()
    }
}

fun main() {
    val problem = 개미굴_14725()
    problem.solution()
}